browser/blueprints.py

```diff
@@ -90,89 +77,64 @@ def upload_file(project_id):
+    if 'frame' in info:
+        del info['frame']
 
+    try:
+        project = Project.get_project(project_id)
         if not project:
             return jsonify({'error': 'project_id not found'}), 404
-
-        state = load_project_state(project)
-        # Perform edit operation on the data file
-        state.action(action_type, info)
-
-        x_changed = state._x_changed
-        y_changed = state._y_changed
-        info_changed = state.info_changed
-
-        state._x_changed = state._y_changed = state.info_changed = False
-
-        # Update object in local database
-        Project.update_project(project, state)
+        edit = get_edit(project)
+        edit.action(action_type, info)
+        payload = edit.make_payload()
+        edit.persist_pickles()  # Must explicitly copy PickleType columns to persist in database
 
     except Exception as e:  # TODO: more error handling to identify problem
         traceback.print_exc()
         return jsonify({'error': str(e)}), 500
 
-    tracks = state.file.readable_tracks if info_changed else False
-
-    if x_changed or y_changed:
-        encode = lambda x: base64.encodebytes(x.read()).decode()
-        img_payload = {}
-
-        if x_changed:
-            raw = state.get_frame(frame, raw=True)
-            img_payload['raw'] = f'data:image/png;base64,{encode(raw)}'
-        if y_changed:
-            img = state.get_frame(frame, raw=False)
-            img_payload['segmented'] = f'data:image/png;base64,{encode(img)}'
-            edit_arr = state.get_array(frame)
-            img_payload['seg_arr'] = edit_arr.tolist()
-
-    else:
-        img_payload = False
-
     current_app.logger.debug('Action "%s" for project "%s" finished in %s s.',
                              action_type, project_id,
                              timeit.default_timer() - start)
 
-    return jsonify({'tracks': tracks, 'imgs': img_payload})
+    return jsonify(payload)
 
 
 @bp.route('/frame/<int:frame>/<int:project_id>')
 def get_frame(frame, project_id):
-    ''' Serve modes of frames as pngs. Send pngs and color mappings of
-        cells to .js file.
-    '''
+    """
+    Serve modes of frames as pngs. Send pngs and color mappings of
+    cells to .js file.
+    """
     start = timeit.default_timer()
-    # Use id to grab appropriate TrackEdit/ZStackEdit object from database
-    project = Project.get_project_by_id(project_id)
-
+    # Get project from database
+    project = Project.get_project(project_id)
     if not project:
         return jsonify({'error': 'project_id not found'}), 404
-
-    state = load_project_state(project)
-
-    # Obtain raw, mask, and edit mode frames
-    img = state.get_frame(frame, raw=False)
-    raw = state.get_frame(frame, raw=True)
-
-    # Obtain color map of the cells
-    edit_arr = state.get_array(frame)
-
+    # Change the frame
+    project.metadata_.frame = frame
+    db.session.commit()
```

Comment:
I made an update method in the Project table that contains the `db.session.commit()` statement. I also replaced the `db.session.commit()` line with `project.update()` in the `persist_pickles` method (which I also renamed to `commit_changes`).

Because I made `update` an instance method, I changed `finish_project` from a static method to an instance method & renamed it to `finish` (so we have `project.finish()` instead of `Project.finish_project(project)`). I also renamed `get_project` to `get` and `create_project` to `create`, but they are both still static methods as they interact with a the Project table, not an instance.